Rework the way USB core stack handles USB Interface Requests. (+reword)

eobalski opened the pull request:
Rework the way how the interfaces are handled by the USB core stack.
Until now USB stack was aware of only the first interface associated
with each class. This led to a problem when not first interface was
addressed by the request.

The usb_cfg_data struct is legacy of initial implementation
of USB device stack. It was implemented to support single function
devices with no composite support. The usb_cfg_data struct was
intended to be used to configure the device by calling
usb_set_config(). Because of how things grow this structure
was used to represent the USB functions rather than USB devices.

This patch:
 - allow to filter out the request at core level.
 - removes device configuration descriptor pointer.
   Adds pointer to list of pointers to USB interface descriptors
   associated with the function it represents.
 - renames some usb_cfg_data structure fields:
    interface -> request_handlers
	Name interface was used here as a communication interface
	between USB core stack and Class implementations.
	This naming could be missleading as USB specification defines
	interface as a part of the USB device descriptor.
    endpoint -> endpoints
	This change was made to reflect that there might be more than one
	endpoint. Additionaly it makes it consistent with newly added
	'interfaces' field.

Fixes: #24200 (USB3CV Chapter 9 tests for ECM)

Commit:
usb: Remove pointer to device descriptor from usb_cfg_data

usb_cfg_data data structure is handling usb funciton specifig data.
USB device descriptor is single instance located in special RAM section.

This deletion is a step forward to make core stack aware of functions
it handles.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Rework the usb core internals.

Rework the way how the interfaces are handled by the USB core stack.
Until now USB stack was aware of only the first interface associated
with each class. This led to a problem when not first interface was
addressed by the request.

This patch allow to filter out the request at core level.

The usb_cfg_data struct is legacy of initial implementation
of USB device stack. Initially the USB device stack was
implemented to support single function devices with
no composite support. This structure was intended to be used to
configure the device in run time by calling usb_set_config().

Because of how things grow the structure was used to represent
the functions instead of devices.

This patch removes device configuration descriptor pointer.
Adds pointer to list of pointers to USB interface descriptors
associated with the function it represents.

This patch should be followed by renaming the structure to something
closer to 'function' specific data as described in USB 2.0 spec.
Additionally interface field should be renamed.
In USB terminology interface is a part of the device function. Here,
it is used as reference to interface between core and class part of
USB device stack.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: ECM fix to pass USB3CV tests.

This patch uses custom request handler to properly
respond to SET/GET interface request.

Fixes #24200

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

Commit:
usb: Rename usb_cfg_data structure fields

Rename USB usb_cfg_data structures to better reflect its meaning.

interface -> request_handlers
    Name interface was used here as a communication interface
    between USB core stack and Class implementations.
    This naming could be missleading as USB specification defines
    interface as a part of the USB device descriptor.

endpoint -> endpoints
    This change was made to reflect that there might be more than one
    endpoint. Additionally it makes it consistent with newly added
    'interfaces' field.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

emob-nordic commented:
I do not insist on last commit which rename things but I think this should be done as well.

jukkar commented on include/usb/usb_device.h:
```diff
@@ -174,8 +174,6 @@ struct usb_interface_cfg_data {
  * may only be updated after calls to usb_deconfig
```
Typo in commit msg `usb: Rework the usb core internals.`
s/devie/device/

emob-nordic replied:
Thanks, will correct this.

jukkar commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -31,6 +31,7 @@ LOG_MODULE_REGISTER(usb_ecm);
 
```
In commit `usb: ECM fix to pass USB3CV tests`
Could we just say `Fixes #24200` instead of the longer string.

emob-nordic replied:
Yes, no problem.

jukkar reviewed (commented):
Very nice commit messages!

emob-nordic commented on subsys/usb/class/netusb/function_ecm.c:
```diff
@@ -345,21 +341,46 @@ static struct netusb_function ecm_function = {
+				return -ENOTSUP;
+			}
+			LOG_DBG("Set interface iface: %d, alt: %d", iface, alt);
+			return -EINVAL;
```
Yes it was but, ekhmm its Zephyr. Let me explain why. We have recently introduced this behaviour in audio PR. `custom_handler` has its special [return type](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/usb/usb_device.h#L160). This is done like that because class implementation handles the status of alternate interface and for set/get request it saves/returns the number of alternate that was requested. For SET however one more thing needs to be done -> proper endpoints associated with that alternate must be enabled or disabled.

Class custom handler is called from [there](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L915). Look closely at this if condition. If I `return 0` there it will return from `usb_handle_standard_request` without calling `usb_handle_std_interface_req` which enabled/disables the associated endpoints.

Reference: [audio](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/audio/audio.c#L736).

anangl replied:
A comment like "This request needs further processing in the core USB stack, hence -EINVAL is returned." would be helpful then. In the audio case, too.

emob-nordic replied:
I will add this as a separate commit.

jfischer-phytec-iot replied:
This behavior is confusing, we have to rework it somehow asap. What about not to do it in custom_handler but implement a specific handler to set interface which would be called from REQ_SET_INTERFACE(usb_set_interface)?

Commit:
usb: Additional comments for usb custom_handler

This commit add additional comments for custom request handlers.
Both ecm and audio classes uses special form of return type to
let the USB core stack handle the same request.

The comment clarifies why in case of a succes -EINVAL is returned.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

anangl commented on subsys/usb/usb_device.c:
```diff
@@ -1197,26 +1198,28 @@ static int class_handler(struct usb_setup_packet *pSetup,
-	struct usb_interface_cfg_data *iface;
+	const struct usb_interface_cfg_data *handlers;
+	const struct usb_cfg_data *cfg = NULL;
+	uint8_t interface_number = sys_le16_to_cpu(pSetup->wIndex);
```
This implicitly casts `uint16_t` to `uint8_t`. I think it would be good to add explicit casting here, to avoid warnings/errors in some future. Same story in [line 1236](https://github.com/zephyrproject-rtos/zephyr/pull/26075/files#diff-609952c961f3c77cb7811bedee4bfd2cR1236).

jfischer-no reviewed (commented):
LGTM for 1c6bd53 and 7232d05. I suggest to drop for now or move to separate PR last tree patches.

emob-nordic commented on subsys/usb/usb_device.c:
```diff
@@ -804,7 +804,10 @@ static bool usb_handle_std_interface_req(struct usb_setup_packet *setup,
 
 	case REQ_GET_INTERFACE:
-		/* there is only one interface, return n-1 (= 0) */
+		/* Interfaces that have alternates are handled by classes code.
```
Tbh I am happy You wrote that. I think I have an idea of how We can accomplish that. Instead of `usb_cfg_data::const usb_if_descriptor *const *const interfaces`

emob-nordic replied:
Tbh I am happy You mentioned that. I think I have an idea of how We can accomplish that.
Instead of `usb_cfg_data :: interfaces` with type of `const struct usb_if_descriptor *const *const` We can create a list of structs which not only points to interface but to alternate (if exists) as well + whatever we want to have there. Having that we will be able to "know" the configured function/class form core.

Simply speaking something like:
`const struct interface_container *const *const` 
Where `struct interface_container` is:
```
struct interface_container {
    struct usb_if_descriptor *iface;
    struct usb_if_descriptor *iface_alt;
    uint8_t curr_alt;
};
```

emob-nordic commented:
>This behavior is confusing, we have to rework it somehow asap. What about not to do it in custom_handler but implement a specific handler to set interface which would be called from REQ_SET_INTERFACE(usb_set_interface)?

So remove it from custom_handler and create another one specific for SET/GET_INTERFACE? This is not simplification but making another hacks in the stack. I am thinking about different approach.

emob-nordic commented:
Rebased on master.

jfischer-phytec-iot commented:
> > This behavior is confusing, we have to rework it somehow asap. What about not to do it in custom_handler but implement a specific handler to set interface which would be called from REQ_SET_INTERFACE(usb_set_interface)?
> 
> So remove it from custom_handler and create another one specific for SET/GET_INTERFACE? This is not simplification but making another hacks in the stack. I am thinking about different approach.

Why adding a dedicated function for required task is a hack?

emob-nordic commented:
> Why adding a dedicated function for required task is a hack?

I named it wrongly. It's not a hack but it makes exception for just one thing. I wish we go in direction of making the stack more compact instead of creating an exception for just one type of request. Imo using `custom_handler` for "problematic" requests is acceptable. I don't have a better idea for it (at least for now). I am not a fan of adding yet another request handler for just one thing.

The solution proposed is fixing an issue we currently have and is presenting the approach of how to deal with such problems. I am thinking of another way to do that so `custom_handler` will not be required at all but its something I need to rethink.

jfischer-phytec-iot commented:
> > Why adding a dedicated function for required task is a hack?
> 
> I named it wrongly. It's not a hack but it makes exception for just one thing. I wish we go in direction of making the stack more compact instead of creating an exception for just one type of request. Imo using `custom_handler` for "problematic" requests is acceptable. I don't have a better idea for it (at least for now). I am not a fan of adding yet another request handler for just one thing.

But it is not a custom_handler. One part of the request has to be handled in the stack self, the other delegated to the class/function. 

> 
> The solution proposed is fixing an issue we currently have and is presenting the approach of how to deal with such problems. I am thinking of another way to do that so `custom_handler` will not be required at all but its something I need to rethink.

Let us rethink together. custom_handler can be removed/deprecated. Please remove the last tree commits so that we can move forward with this PR.

emob-nordic commented:
> Let us rethink together. custom_handler can be removed/deprecated. Please remove the last tree commits so that we can move forward with this PR.

I think this idea
> Simply speaking something like:
`usb_cfg_data :: const struct interface_container *const *const`
Where struct interface_container is:

```
struct interface_container {
    struct usb_if_descriptor *iface;
    struct usb_if_descriptor *iface_alt;
    uint8_t curr_alt;
};
```

Is a better solution. I consider closing this PR and reopening as Stable API change Issue + PR.

emob-nordic commented:
Agreed to follow up the rework on separate [topic-usb](https://github.com/zephyrproject-rtos/zephyr/tree/topic-usb) branch.
